### PR TITLE
feature(tailwindcss-transformer,ui): Scaffold packages

### PR DESCRIPTION
**Pull request by alexcarpenter**
## Description

In this PR we introduce two new private _experimental_ packages.

## `tailwindcss-transformer`

The goal of this package is to enable authoring components using Tailwind CSS and exporting a component with hashed classNames with accompanying stylesheet.

#### Input:

```jsx
export function Example({ flag }) {
  let className = cn('absolute inset-0', flag && 'uppercase');
  return <div className={cn('flex items-center text-sm', className)} />;
}
```

#### Output:

```jsx
export function Example({ flag }) {
  let className = cn('cl-7601190e', flag && 'cl-d2cf63c7');
  return <div className={cn('cl-f64ae6a6', className)} />;
}
```

```css
.cl-7601190e {
  position: absolute;
  inset: 0;
}

.cl-d2cf63c7 {
  text-transform: uppercase;
}

.cl-f64ae6a6 {
  display: flex;
  align-items: center;
  font-size: 0.875rem;
}
```

### Utilities
- `transform` - transforms a React component with TW classes into hashed classNames
- `generateStylesheet` - generates a single stylesheet based on classNames extracted from `transform` utility

## `ui`

The `ui` package is where we will be rebuilding the AIO components. We run the components through tsup with the `tailwindcss-transformer` as an esbuild plugin.

## Checklist

- [ ] `npm test` runs as expected.
- [ ] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [x] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

### CODE REVIEW

File: packages/ui/example/app/globals.css [hunk 1]
```diff
@@ -0,0 +1,3 @@
+@tailwind base;
+@tailwind components;
+@tailwind utilities;
```
Comment (kylemac) [inline packages/ui/example/app/globals.css:3]: i'm wondering if we should intentionally work in an example that doesn't include tailwind - to ensure there's no leaking across from AIO components and the next-app

**Comment by panteliselef** [comment]
Will be used to move the ui components out of clerk-js, or this is intended for a complete rewrite ?

**Comment by kylemac** [comment]
@panteliselef no firm decisions on where files will live long term (wasn't sure if that was part of your question or not) - but the plan is to do a complete rebuild using elements under the hood. that said, this is all experimental/poc atm

File: packages/ui/example/app/globals.css (hunk 1 above)
Comment (alexcarpenter) [reply packages/ui/example/app/globals.css]: Removed the tailwind exports.

File: packages/ui/package.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  "name": "@clerk/ui",
+  "version": "0.1.0",
+  "private": true,
+  "license": "MIT",
+  "author": {
+    "name": "Clerk, Inc.",
+    "email": "support@clerk.com",
+    "url": "git+https://github.com/clerk/javascript.git"
+  },
+  "main": "index.js",
```
Comment (LekoArts) [inline packages/ui/package.json:11]: Needs to be replace with an exports map

**Review by LekoArts — CHANGES_REQUESTED** [review]

File: packages/tailwindcss-transformer/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+) {
+  let stylesheet = '@tailwind base;\n';
+
+  stylesheet += ctx.globalCss || '';
+
+  for (const [cn, value] of styleCache) {
+    stylesheet += `html :where(.${cn}) { @apply ${value} }\n`;
+  }
+
+  // @ts-ignore
```
[the remark this answers is not in the document]
Comment (alexcarpenter) [reply packages/tailwindcss-transformer/src/index.ts]: Fixed the issue 👍🏼

File: packages/tailwindcss-transformer/tests/__snapshots__/generateStylesheet.test.js.snap
[the remark this answers is not in the document]
Comment (alexcarpenter) [reply packages/tailwindcss-transformer/tests/__snapshots__/generateStylesheet.test.js.snap:1]: Yeah, there will likely be a step added in the future to minify the CSS. Will hold off for now though.

File: packages/tailwindcss-transformer/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+  return true;
+}
+
+function generateHashedClassName(value: string) {
+  return 'cl-' + createHash('sha256').update(value, 'utf8').digest('hex').slice(0, 8);
+}
+
+let clRegex = /^cl-[a-z0-9]{8}$/;
+
+export async function transform(code: string, ctx: { styleCache: StyleCache }) {
```
[the remark this answers is not in the document]
Comment (alexcarpenter) [reply packages/tailwindcss-transformer/src/index.ts]: > const Button = ({ children }) => {
  const myStyle = 'flex'
  return <button className={myStyle}>{children}</button>
}

I will add a test case for this, I don't anticipate we'd use styling in this manner FWIW.

> const ButtonPrimitive = ({ children, style }) => <button className={style}>{children}</button>
>
> const Button =  ({ children }) => <ButtonPrimitive style="flex">{children}</ButtonPrimitive>

We wouldn't support this syntax.

[the remark this answers is not in the document]
Comment (LekoArts) [reply packages/tailwindcss-transformer/src/index.ts]: Then let's add these limitations to the README so it's obvious what won't work 👍

**Review by BRKalow — APPROVED** [review]
🚢

**Pull request merged**